libsolidity/interface/CompilerStack.cpp

```diff
@@ -808,15 +808,19 @@ evmasm::AssemblyItems const* CompilerStack::assemblyItems(std::string const& _co
 	solAssert(m_stackState == CompilationSuccessful, "Compilation was not successful.");
 
 	Contract const& currentContract = contract(_contractName);
-	return currentContract.evmAssembly ? &currentContract.evmAssembly->items() : nullptr;
+	if (currentContract.evmAssembly)
+		solAssert(currentContract.evmAssembly->codeSections().size() == 1, "Expected a single code section in legacy codegen.");
+	return currentContract.evmAssembly ? &currentContract.evmAssembly->codeSections().front().items : nullptr;
```

Comment:
Actually, can we even be sure that we're dealing with legacy here? Unless we have a general check against somewhere in `StandardCompiler` and `CommandLineInterface` that disables some outputs on EOF, it's actually reachable by the user.

So in addition to the assert in the legacy case (which should never fail) this also needs a validation against EOF (which may fail and should be handled gracefully):

```suggestion
	if (!currentContract.evmAssembly)
	    return nullptr;

	solUnimplementedAssert(!m_eofVersion.has_value(), "EVM assembly output not implemented for EOF yet.");
	solAssert(currentContract.evmAssembly->codeSections().size() == 1, "Expected a single code section in legacy codegen.");
	return &currentContract.evmAssembly->codeSections().front().items;
```

And again, `solUnimplementedAssert()` is not really an assert just a badly named macro :) It throws `UnimplementedFeatureError`, which we catch and display as a normal error.

Same for `runtimeAssemblyItems()` and (probably) `YulStack::assembleWithDeployed()`.

Reply:
Done

Reply:
There's still `YulStack` and `runtimeAssemblyItems()` left :)

Reply:
Fixed. :) Not fully sure if these are places where it should be checked.

Reply:
Well, you could check by trying to run `solc` with `--asm` and `--eof-version` and giving it some code that would run into this check. I'm almost certain that in those `CompilerStack` you can and I suspect that this is the case with `YulStack` as well. Actually, I see we don't have `--asm-runtime` so maybe you can't actually trigger the one in `runtimeAssemblyItems()`, but I think the check should still be there.

But if you're unsure, it's also perfectly fine to add them just in case, like I see you did. Having `solUnimplementedAssert()` in more places than strictly necessary won't break anything and is not incorrect.